### PR TITLE
Fix exception caused by `ToolStripMenuItem.DropDownClosed` hiding itself.

### LINKED ISSUE #7884: Hiding a ToolStripMenuItem on DropDownClosed throws exception if closed by hovering over different menu item
### .NET version

Version:   6.0.300
Commit:    8473146e7d

Also did not work in .NET Framework 4.8

### Did it work in .NET Framework?

No

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

No, it did not work in any of the earlier releases of .NET Core or .NET 5+.

### Issue description

<details>
<summary><b>Exception stack trace</b></summary>
<pre><code>System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
   at System.Collections.ArrayList.get_Item(Int32 index)
   at System.Windows.Forms.ToolStripItemCollection.get_Item(Int32 index)
   at System.Windows.Forms.ToolStrip.ClearAllSelectionsExcept(ToolStripItem item)
   at System.Windows.Forms.ToolStrip.NotifySelectionChange(ToolStripItem item)
   at System.Windows.Forms.ToolStripItem.Select()
   at System.Windows.Forms.ToolStripItem.HandleMouseEnter(EventArgs e)
   at System.Windows.Forms.ToolStrip.OnMouseMove(MouseEventArgs mea)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ScrollableControl.WndProc(Message& m)
   at System.Windows.Forms.ToolStrip.WndProc(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, WM msg, IntPtr wparam, IntPtr lparam)</code></pre>
</details>

<details>
<summary><b>Screen recording of the issue</b></summary>

https://user-images.githubusercontent.com/23175424/193817496-069210c5-edd2-4014-903f-e68a3e9e67d1.mp4

</details>

[Test VS solution to reproduce (includes .NET 6 and .NET Framework 4.8 projects)](https://github.com/jonaskohl/MenuStripBug)

---

Problem is probably caused by a race condition in [`ToolStrip.ClearAllSelectionsExcept(ToolStripItem)`](https://github.com/dotnet/winforms/blob/main/src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs#L1951-L2020).

### Steps to reproduce

1. Have a `MenuStrip` with at least one visible and one hidden `ToolStripMenuItem`. Have an event listener for `DropDownClosed` on the hidden item which sets that item's `Visible` property to `false`.
2. Programatically trigger the hidden menu item by first setting `menuItem.Visible = true` and then calling `menuItem.ShowDropDown()`.
3. Hover over another menu item, thus closing the (intially) hidden item's dropdown.
4. A `System.ArgumentOutOfRangeException` will be thrown.

### PR DESCRIPTION
**Pull request by elachlan**
Fixes #7884

First commit contains the test to prove the failure. Thus proving the subsequent fix actually fixes it.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/8454)

### CODE REVIEW

**Comment by elachlan** [comment]
Test is failing, so now I will include the fix.

File: src/System.Windows.Forms/tests/IntegrationTests/UIIntegrationTests/ToolStripTests.cs [hunk 1]
```diff
@@ -61,5 +61,31 @@ static void ShowForm(ImageList sharedImageList)
...
+        public void ToolStrip_Hiding_ToolStripMenuItem_OnDropDownClosed_ShouldNotThrow()
+        {
+            using Form form = new();
+
+            using ToolStripMenuItem menu1 = new("Menu1");
+            using ToolStripMenuItem menu1Item1 = new("Item1");
+            menu1.DropDownItems.Add(menu1Item1);
+
+            using ToolStripMenuItem hiddenMenu = new("hiddenMenu");
+            using ToolStripMenuItem menu2Item1 = new("Item1");
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/IntegrationTests/UIIntegrationTests/ToolStripTests.cs:76]: Suggestion: `hiddenMenuItem1 `

**Comment by Tanya-Solyanik** [comment]
What was the reason for the unhandled exception box instead of the test failure?

**Comment by elachlan** [comment]
> What was the reason for the unhandled exception box instead of the test failure?

See #8455

**Review by Tanya-Solyanik — APPROVED** [review]
Looks good!

**Review by dmitrii-drobotov — APPROVED** [review]
LGTM

**Pull request merged**